### PR TITLE
fix(core): prevents findPackageLocation from being called from unactivated linkers

### LINKED ISSUE #3777: [Bug?]: Ambiguous error message for running `yarn` in nested project
### Self-service

- [ ] I'd be willing to implement a fix

### Describe the bug

Ambiguous error message for running `yarn` in nested project

### To reproduce

```sh
~> mkdir /tmp/a
~> cd /tmp/a
/tmp/a> yarn init -2
/tmp/a> mkdir b
/tmp/a> cd b
/tmp/a/b> npm init -y
/tmp/a/b> yarn
Usage Error: The project in /tmp/a/package.json doesn't seem to have been installed - running an install there might help

$ yarn install [--json] [--immutable] [--immutable-cache] [--check-cache] [--inline-builds] [--mode #0]
```

### Environment

```shell
System:
    OS: Linux 5.13 Manjaro Linux
    CPU: N/A
  Binaries:
    Node: 16.11.1 - /tmp/xfs-55aacde0/node
    Yarn: 3.1.0 - /tmp/xfs-55aacde0/yarn
    npm: 8.1.3 - /usr/bin/npm
```

### Additional context

The following error message is useful:
```
/tmp/a/b> yarn init -2
Usage Error: Cannot use the --install flag from within a project subdirectory

$ yarn init [-p,--private] [-w,--workspace] [-i,--install]
```

Ideally this error message should be used instead of the current error message (running an install there might help).

### PR DESCRIPTION
**Pull request by arcanis**
**What's the problem this PR addresses?**

The `findLocatorForLocation` method was attempting to get the locator from the linkers, then to query the package location (in order to get the root of the package). The problem was that it was doing this even on disabled linkers, causing crashes in the `findPackageLocation` calls since their install states were missing. As a result, the errors reported by running `yarn install` within a non-workspace directory were incorrect.

Closes #3777
Closes #3792 (Supersedes)

**How did you fix it?**

This diff ensures that the linkers never advertise anything as a package they own when they are disabled.

I think it should be a fairly safe change, but I'll keep an eye on the tests in case I forgot an edge case.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: packages/plugin-nm/sources/NodeModulesLinker.ts [hunk 1]
```diff
@@ -35,10 +35,13 @@ export class NodeModulesLinker implements Linker {
   private installStateCache: Map<string, Promise<InstallState | null>> = new Map();
 
   supportsPackage(pkg: Package, opts: MinimalLinkOptions) {
-    return opts.project.configuration.get(`nodeLinker`) === `node-modules`;
+    return this.isEnabled(opts);
   }
 
   async findPackageLocation(locator: Locator, opts: LinkOptions) {
+    if (!this.isEnabled(opts))
+      throw new Error(`Assertion failed: Expected the npm linker to be enabled`);
```
Comment (paul-soporan) [inline packages/plugin-nm/sources/NodeModulesLinker.ts:43]:
```suggestion
      throw new Error(`Assertion failed: Expected the node-modules linker to be enabled`);
```

File: packages/plugin-pnpm/sources/PnpmLinker.ts [hunk 2]
```diff
@@ -15,6 +15,9 @@ export class PnpmLinker implements Linker {
   }
 
   async findPackageLocation(locator: Locator, opts: LinkOptions) {
+    if (opts.project.configuration.get(`nodeLinker`) !== `pnpm`)
```
Comment (paul-soporan) [inline packages/plugin-pnpm/sources/PnpmLinker.ts:18]: Wouldn't it be better to have an `isEnabled` method here too for consistency?

File: packages/plugin-pnpm/sources/PnpmLinker.ts [hunk 3]
```diff
@@ -229,19 +235,18 @@ class PnpmInstaller implements Installer {
...
-                if (entry === Filename.nodeModules) {
-                  const extraneous = await getNodeModulesListing(p);
-                  extraneous.delete(identComponents.join(ppath.sep) as PortablePath);
-                  return cleanNodeModules(p, extraneous);
-                } else {
-                  return xfs.removePromise(p);
-                }
-              }));
-            })
+          removals.push(xfs.readdirPromise(storeEntryPath).then(entries => {
```
Comment (paul-soporan) [inline packages/plugin-pnpm/sources/PnpmLinker.ts:238]: Shouldn't the `.catch` at least be directly next to the closing paren if the `.then` got moved to the same line as the async function call?

Comment (arcanis) [reply packages/plugin-pnpm/sources/PnpmLinker.ts:238]: Yes, I don't know why this code changed, I didn't do it 🤔
